WcaOnRails/app/javascript/edit-schedule/SchedulesEditor/CustomActivity.jsx

```diff
@@ -0,0 +1,114 @@
+import React from 'react'
+import { Button, Modal } from 'react-bootstrap'
+import {
+  addActivityToCalendar,
+  eventModifiedInCalendar,
+  momentToIso,
+} from './calendar-utils'
+import { scheduleElementSelector } from './fullcalendar'
+
+export const commonActivityCodes = {
+  "other-registration": "Registration",
+  "other-breakfast": "Breakfast",
+  "other-lunch": "Lunch",
+  "other-dinner": "Dinner",
+  "other-awards": "Awards",
+  "other-misc": "Other",
+};
+
+export const modeDetails = {
+  create: {
+    modalTitle: "Add a custom activity",
+    buttonText: "Add",
+    action: (hide, eventData) => {
+      eventData.startTime = momentToIso(eventData.start);
+      eventData.endTime = momentToIso(eventData.end);
+      addActivityToCalendar(eventData);
+      hide();
+    },
+  },
+  edit: {
+    modalTitle: "Edit activity",
+    buttonText: "Save",
+    action: (hide, eventData) => {
+      // CustomActivityModal only edit name, fix the title to enable a simple update
+      eventData.title = eventData.name;
+      eventModifiedInCalendar(eventData);
+      $(scheduleElementSelector).fullCalendar("updateEvent", eventData);
+      hide();
+    },
+  },
+};
+
+export class CustomActivityModal extends React.Component {
+
+  componentWillMount() {
+    this.setState({
+      ...this.props.eventProps
+    });
+  }
+
+  componentWillReceiveProps(newProps) {
+    this.setState({
+      ...newProps.eventProps
+    });
+  }
+
+  render () {
+    let { show, handleHideModal, actionDetails, eventProps } = this.props;
+    let { modalTitle, buttonText, action } = actionDetails;
+    let timeText = "No time selected";
+    if (eventProps.start && eventProps.end) {
+      timeText = `On ${eventProps.start.format("dddd, MMMM Do YYYY")}, from ${eventProps.start.format("H:mm")} to ${eventProps.end.format("H:mm")}.`;
+    }
+
+    let handlePropChange = (propName, e) => {
+      let newState = {};
+      newState[propName] = e.target.value;
+      if (propName == "activityCode") {
+        // On change of activity code, we can update the activity name to the default
+        newState.name = commonActivityCodes[newState.activityCode];
+      }
+      this.setState(newState);
+    };
```

Comment:
Given that there are only two inputs and they are handled differently (the `if` statement), I think it makes more sense to have two separate handlers
```js
let handleNameChange = event => {
  this.setState({ name: event.target.value });
};

let handleActivityCodeChange = event => {
  this.setState({
    activityCode: event.target.value,
    name: commonActivityCodes[event.target.value]
  });
};
```
Another advantage of this is that we don't create new functions on each render, which is the case when we use arrow functions like `onChange={e => handlePropChange("activityCode", e)}`.